### ISSUE TITLE
AR cmphet scoring broken for multi-sample VCFs

### ISSUE THREAD

**Issue by damiansm** [exomiser/Exomiser#176]
Current scoring algorithm for all versions just takes the average of the top 2 scoring hets for a gene that have passed the compatibleWithAR check but these two variants are not necessarily compatible together for a multi-sample VCF e.g. these two were identified as the top scoring hets in a Genomiser run but the pedigree is an unaffected mother (middle genotype) and two affected brothers:

1 	98518687 	T 	A 	733 	PASS 	0/1:0/1:0/1 	0 	regulatory_region_variant 	? 	DPYD 	. 	. 	. 	. 	0.953466 	rs61786698 	0 	. 	. 	. 	. 	. 	. 	. 	. 	. 	. 	0.953466 	0.80476695 	0.952971 	0.9787295
1 	98518683 	T 	A 	733 	PASS 	0/1:0/1:0/1 	0 	regulatory_region_variant 	? 	DPYD 	. 	. 	. 	. 	0.952476 	rs5006621 	0 	. 	.

This becomes a big issue for Genomiser runs where lots of non-coding variants pass the variant filters.

On development the following code fixes the issue but we need to implement this in a nicer way and see if there is a way of doing this without passing the pedigree to RawGeneScorer:

private float calculateAutosomalRecessiveFilterScore(List<VariantEvaluation> variantEvaluations, int sampleId, Pedigree pedigree) {

    // THIS LOGIC IS WRONG FOR MULTI-SAMPLE VCFs AS THE TOP TWO HET VARIANTS MAY NOT BE CONSISTENT WITH AR
    // i.e. THEY COULD BE HET IN UNAFFECTED AS WELL
    List<VariantEvaluation> hetContributingVariants = new ArrayList<>();
    double bestCompHetScore = 0;
    InheritanceCompatibilityChecker inheritanceCompatibilityChecker = new InheritanceCompatibilityChecker.Builder().pedigree(pedigree).addMode(ModeOfInheritance.AUTOSOMAL_RECESSIVE).build();
    for (VariantEvaluation ve1: variantEvaluations){
        VariantContext va1 = ve1.getVariantContext();
        for (VariantEvaluation ve2: variantEvaluations){
            if (ve1.equals(ve2))
                    continue;
            VariantContext va2 = ve2.getVariantContext();
            List<VariantContext> geneVariants = new ArrayList<VariantContext>();
            geneVariants.add(va1);
            geneVariants.add(va2);
            try {
                List<VariantContext> compatibleVariants = inheritanceCompatibilityChecker.getCompatibleWith(geneVariants);
                if (compatibleVariants.size() == 2){
                    //logger.info("Compatible Variant 1 " + ve1.toString());
                    //logger.info("Compatible Variant 2 " + ve1.toString());
                    double compHetScore = (ve1.getVariantScore() + ve2.getVariantScore())/2;
                    if (compHetScore > bestCompHetScore) {
                        bestCompHetScore = compHetScore;
                        hetContributingVariants = new ArrayList<>();
                        hetContributingVariants.add(ve1);
                        hetContributingVariants.add(ve2);
                    }
                    //logger.info("Score is " + compHetScore + " + and best score is " + bestCompHetScore);
                }
            }
            catch (Exception ex) {
                logger.error(null, ex);
            }
        }
    }

    /*List<VariantEvaluation> heterozygous = variantEvaluations.stream()
            .filter(variantIsHeterozygous(sampleId))
            .sorted(Comparator.comparing(VariantEvaluation::getVariantScore).reversed())
            .limit(2)
            .collect(toList());*/

    Optional<VariantEvaluation> bestHomozygousAlt = variantEvaluations.stream()
            .filter(variantIsHomozygousAlt(sampleId))
            .max(Comparator.comparing(VariantEvaluation::getVariantScore));

    // Realised original logic allows a comphet to be calculated between a top scoring het and second place hom which is wrong
    // Jannovar seems to currently be allowing hom_ref variants through so skip these as well
    /*double bestCompHetScore = heterozygous.stream()
            .mapToDouble(VariantEvaluation::getVariantScore)
            .average()
            .orElse(0f);*/

    double bestHomAltScore = bestHomozygousAlt
            .map(VariantEvaluation::getVariantScore)
            .orElse(0f);

    double bestScore = Double.max(bestHomAltScore, bestCompHetScore);

    if (BigDecimal.valueOf(bestScore).equals(BigDecimal.valueOf(bestCompHetScore))) {
        //heterozygous.forEach(VariantEvaluation::setAsContributingToGeneScore);
        for (VariantEvaluation contributingVariant : hetContributingVariants){
            contributingVariant.setAsContributingToGeneScore();
        }
    } else {
        bestHomozygousAlt.ifPresent(VariantEvaluation::setAsContributingToGeneScore);
    }

    return (float) bestScore;
}

Need to fix in development/master and probably release a patch to 7.2.3 as well

**Comment by julesjacobsen** [comment]
This issue is related to #147 in that we need to define a proband in order to calculate the inheritance mode.

**Comment by damiansm** [comment]
@julesjacobsen Did we ever put the code fix above into development or anywhere else. Does not look like it?

**Comment by pnrobinson** [comment]
@julesjacobsen Would you have time to collaborate on fixing some issues with the pedigree filter in Jannovar/Exomiser? There is one bug that will only affect very rare multiple sample VCF files, but I want to write some JUnit code for this situation and consider how to address the code.

**Comment by damiansm** [comment]
NB If run with GEL local frequency then both example variants in LP2000254_H05 get filtered out as 8% and 21% respectively

**Comment by julesjacobsen** [comment]
@pnrobinson maybe. It would be good to get these things fixed in Jannovar rather than having to patch things up in Exomiser afterwards. Can we talk about this tomorrow?

**Comment by pnrobinson** [comment]
@julesjacobsen definitely,I meant that we should do this in Jannovar. I would like to start by making some JUnit tests that would represent the requirements of the situation, and thus we might want to create some PED files and a few small VCF files with examples of the situations that seem not to be working at present. I'll send an email about this, currently in transit

**Comment by damiansm** [comment]
I have done some more digging and Jannovar and Exomiser are doing the right thing on the current dev branch but we have just some missing logic in our scoring algorithm.

A VCF containing just 
1       98518687        .       T       A       733     SC;alleleBias   BRF=0.11;FR=0.3333;HP=2;HapScore=3;MGOF=51;MMLQ=16;MQ=56.56;NF=15;NR=5;PP=733;QD=41.3234;SC=TGTGTGTGTGTGTGTGTGTGT;SbPval=0.29;Source=Platypus;TC=78;TCF=38;TCR=40;TR=20;WE=98518701;WS=98518671;OLD_CLUMPED=1:98518683:TGTGTGTGTGT/AGAGAGAGAGA  GT:GL:GOF:GQ:NR:NV:PS   1|0:-45.4,0,-74.58:30:99:22:1:98518683  1|0:-66.65,0,-44.48:51:99:32:5:98518683 1|0:-26.07,0,-34.97:42:99:24:14:98518683
1       98518683        .       T       A       733     SC;alleleBias   BRF=0.11;FR=0.3333;HP=2;HapScore=3;MGOF=51;MMLQ=16;MQ=56.56;NF=15;NR=5;PP=733;QD=41.3234;SC=TGTGTGTGTGTGTGTGTGTGT;SbPval=0.29;Source=Platypus;TC=78;TCF=38;TCR=40;TR=20;WE=98518701;WS=98518671;OLD_CLUMPED=1:98518683:TGTGTGTGTGT/AGAGAGAGAGA  GT:GL:GOF:GQ:NR:NV:PS   1|0:-45.4,0,-74.58:30:99:22:1:98518683  1|0:-66.65,0,-44.48:51:99:32:5:98518683 1|0:-26.07,0,-34.97:42:99:24:14:98518683

correctly fails the Jannovar inheritance filter with the supplied pedigree as the middle genotype is for the unaffected mother.

HOWEVER, if you also add:

1       97723020        .       A       G       1141    SC;QD   BRF=0.06;FR=0.3333;HP=2;HapScore=2;MGOF=37;MMLQ=36;MQ=58.47;NF=72;NR=46;PP=1141;QD=9.96416;SC=GAGAGAGAGAAAGAGAGAGAG;SbPval=0.6;Source=Platypus;TC=128;TCF=77;TCR=51;TR=118;WE=97723028;WS=97722998      GT:GL:GOF:GQ:NR:NV      1/0:-51.62,0,-5.3:22:53:40:33   0/0:0,-2.07,-57.46:37:21:38:36  1/0:-66.78,0,-3.49:23:35:50:49

then all 3 variants pass the inheritance filter and are set as passed at the variant level as either variant 1 and 3 or variant 2 and 3 can account for the cmphet scenario.

I also confirmed that the code removes any other DPYD variants that are not compatible with AR i.e. it does not just keep all the previously passed variants for a gene that has some compatible variants.

The issue remains that variants 1 and 2 have the best scores and are used to calculate the geneVariantScore, despite the pair not being compatible with AR.

Not sure of the solution but would probably require the pedigree being available at the scoring stage and a second call to Jannovar to see if the pair were compatible

**Comment by damiansm** [comment]
I confirmed again that the following fixes the issue in RawGeneScorer but is inelegant and involves passing the pedigree object down to it. Any better ideas @julesjacobsen 

   ```
 private float calculateAutosomalRecessiveFilterScore(List<VariantEvaluation> variantEvaluations, int sampleId, Pedigree pedigree) {

        List<VariantEvaluation> hetContributingVariants = new ArrayList<>();
        double bestCompHetScore = 0;

        VariantContextMendelianAnnotator inheritanceAnnotator = new VariantContextMendelianAnnotator(pedigree,false,false);

        //InheritanceCompatibilityChecker inheritanceCompatibilityChecker = new InheritanceCompatibilityChecker.Builder().pedigree(pedigree).addMode(ModeOfInheritance.AUTOSOMAL_RECESSIVE).build();
        for (VariantEvaluation ve1: variantEvaluations){
            VariantContext va1 = ve1.getVariantContext();
            for (VariantEvaluation ve2: variantEvaluations){
                if (ve1.equals(ve2))
                    continue;
                VariantContext va2 = ve2.getVariantContext();
                List<VariantContext> geneVariants = new ArrayList<VariantContext>();
                geneVariants.add(va1);
                geneVariants.add(va2);
                try {
                    //List<VariantContext> compatibleVariants = inheritanceCompatibilityChecker.getCompatibleWith(geneVariants);

                    ImmutableMap<ModeOfInheritance, ImmutableList<VariantContext>> compatibleMap = inheritanceAnnotator.computeCompatibleInheritanceModes(geneVariants);
                    if (compatibleMap.get(ModeOfInheritance.AUTOSOMAL_RECESSIVE).size() == 2){
                        double compHetScore = (ve1.getVariantScore() + ve2.getVariantScore())/2;
                        if (compHetScore > bestCompHetScore) {
                            bestCompHetScore = compHetScore;
                            hetContributingVariants = new ArrayList<>();
                            hetContributingVariants.add(ve1);
                            hetContributingVariants.add(ve2);
                        }
                    }
                }
                catch (Exception ex) {
                    logger.error(null, ex);
                }
            }
        }


        /*List<VariantEvaluation> heterozygous = variantEvaluations.stream()
                .filter(variantIsHeterozygous(sampleId))
                .sorted(Comparator.comparing(VariantEvaluation::getVariantScore).reversed())
                .limit(2)
                .collect(toList());*/

        Optional<VariantEvaluation> bestHomozygousAlt = variantEvaluations.stream()
                .filter(variantIsHomozygousAlt(sampleId))
                .max(Comparator.comparing(VariantEvaluation::getVariantScore));

        // Realised original logic allows a comphet to be calculated between a top scoring het and second place hom which is wrong
        // Jannovar seems to currently be allowing hom_ref variants through so skip these as well
        /*double bestCompHetScore = heterozygous.stream()
                .mapToDouble(VariantEvaluation::getVariantScore)
                .average()
                .orElse(0f);*/

        double bestHomAltScore = bestHomozygousAlt
                .map(VariantEvaluation::getVariantScore)
                .orElse(0f);

        double bestScore = Double.max(bestHomAltScore, bestCompHetScore);

        if (BigDecimal.valueOf(bestScore).equals(BigDecimal.valueOf(bestCompHetScore))) {
            //heterozygous.forEach(VariantEvaluation::setAsContributingToGeneScore);
            for (VariantEvaluation contributingVariant : hetContributingVariants){
                contributingVariant.setAsContributingToGeneScore();
            }
        } else {
            bestHomozygousAlt.ifPresent(VariantEvaluation::setAsContributingToGeneScore);
        }

        return (float) bestScore;
    }
```

**Comment by julesjacobsen** [comment]
Good findings! The Jannovar tests only even consider a maximum of two variants at a time. With comp het you actually need to know the other variants with which a given variant is compatible which cannot be conveyed with a simple flag on all the passed variants.

As you say, I think the only solution to this is to check again at the scoring stage to generate a list of compatible comp het pairs and compare their scores against hom alt variants for recessive modes.

I'll get onto it.

**Comment by damiansm** [comment]
Yes - think turning my code into a more elegant stream but sticking to passing the pedigree down to the scorer is the only way to go. Cmphets ruin everything in RD analysis!

**Comment by julesjacobsen** [comment]
I think I'll look at moving the comp het pairing into the ```InheritanceModeAnalyser ``` and the keep the scoring of these in the ```GeneScorer```. This will still require the Pedigree being passed into the ```GeneScorer```, but that's not the end of the world.

**Comment by julesjacobsen** [comment]
Post fix this is what comes out:
```
1	97849802	G	GAA	540.0	PASS	1/1:0/1:1/1	0	coding_transcript_intron_variant	DPYD:uc001drv.3:c.1906-1786_1906-1785insTT:p.(=)	DPYD	.	.	.	.	0.484159	.	0.0	.	.	.	.	.	.	.	.	.	.	0.484159	0.71613175	0.484159	0.19815478	CONTRIBUTING_VARIANT
```

Is this what you got from your code?

**Comment by damiansm** [comment]
I got:
`1	97799095	G	T	87.0	PASS	0/1:0/0:0/1	0	coding_transcript_intron_variant	DPYD:uc001drv.3:c.2059-27242C>A:p.(=)	DPYD	.	.	.	.	0.242575	.	0.0	.	.	.	.	.	0.242575	0.80476695	0.5980205	0.6385635	CONTRIBUTING_VARIANT
1	98518687	T	A	733.0	PASS	0/1:0/1:0/1	0	regulatory_region_variant	?	DPYD	.	.	.	.	0.953466	rs61786698	0.0	.	.	.	.	.	.	.	.	.	0.953466	0.80476695	0.5980205	0.6385635	CONTRIBUTING_VARIANT
`

i.e. I got a better scoring cmphet match I think. My VCF was whole of chr1 I think.
